File: common/src/main/java/dev/obscuria/fragmentum/util/event/Event.java
package dev.obscuria.fragmentum.util.event;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

@SuppressWarnings("unused")
public final class Event<T>
{
    private final List<Registration<T>> registrations = new ArrayList<>();
    private boolean dirty = false;

    public EventToken register(T listener)
    {
        return register(1000, listener);
    }

    public EventToken register(int priority, T listener)
    {
        final var token = new Token(UUID.randomUUID(), this);
        registrations.add(new Registration<>(token, priority, listener));
        dirty = true;
        return token;
    }

    public void unregister(EventToken token)
    {
        registrations.removeIf(it -> it.token == token);
    }

    public void broadcast(EventHandler<T> handler)
    {
        this.sortIfDirty();
        registrations.forEach(it -> handler.handle(it.listener));
    }

    private void sortIfDirty()
    {
        if (!dirty) return;
        Collections.sort(registrations);
        this.dirty = false;
    }

    private record Registration<T>(EventToken token, int priority, T listener) implements Comparable<Registration<T>>
    {
        @Override
        public int compareTo(Registration<T> other)
        {
            return Integer.compare(priority, other.priority);
        }
    }

    private record Token(UUID uuid, Event<?> event) implements EventToken
    {
        @Override
        public void unregister()
        {
            event.unregister(this);
        }
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/event/Event.kt
package dev.obscuria.fragmentum.tools.event

import java.util.*

class Event<T> {

    private val registrations: MutableList<Registration<T>> = mutableListOf()
    private var dirty = false

    infix fun register(listener: T): EventToken {
        return register(1000, listener)
    }

    fun register(priority: Int, listener: T): EventToken {
        val token = Token(UUID.randomUUID(), this)
        registrations.add(Registration(token, priority, listener))
        this.dirty = true
        return token
    }

    infix fun unregister(token: EventToken) {
        registrations.removeIf { it.token == token }
    }

    infix fun broadcast(handler: EventHandler<T>) {
        this.sortIfDirty()
        registrations.forEach { handler.handle(it.listener) }
    }

    private fun sortIfDirty() {
        if (!dirty) return
        registrations.sortWith(Registration<T>::compareTo)
        this.dirty = false
    }

    private class Registration<T>(
        val token: EventToken,
        val priority: Int,
        val listener: T
    ) : Comparable<Registration<T>> {

        override fun compareTo(other: Registration<T>): Int {
            return this.priority.compareTo(other.priority)
        }
    }

    data class Token(val uuid: UUID, val event: Event<*>) : EventToken
    {
        override fun unregister()
        {
            event.unregister(this)
        }
    }
}